### PR TITLE
Add option in Environment to handle stripped nulls in payload

### PR DESCRIPTION
**Pull request by leoasis**
Adding the feature to handle stripped nulls in responses as an environment option, passing it through to the query executor and then to the response normalizer, which actually does the job of writing the fields with nulls or not based on the flag.

In the QueryExecutor, I explicitly passed the value to `false` when processing optimistic responses. I noticed that this path is not only used for optimistic responses defined in cases such as `commitMutation`, but also executed when the server may send an optimistic response using an extension. I think this case should still be stripping nulls, as those are coming from the server (if I understood correctly). Would love to confirm this and change the code so that I take into account that case (and change the test I added).

Also pass the value through for all the other async cases, such as defer, streams, module imports. I'm not very familiar with how those work and how they should play along with this option, so I'd appreciate some feedback. Based on [this comment](https://github.com/facebook/relay/issues/3052#issuecomment-602788832), it seems that based on how they are traversed it should not be an issue.

Finally, if this approach seems to be the way to go, happy to work on adding documentation for this (would love pointers on where would be the best place to put this).

Fixes https://github.com/facebook/relay/issues/3052

### CODE REVIEW

**Comment by maraisr** [comment]
Just my 2 cents; I feel it should rather be "markMissingAsNull" - strippedNulls? What even is that?

**Comment by leoasis** [comment]
Open to any suggestion for the naming! Just to add to the "stripped nulls" rationale, that's what the response looks like: it's a regular GraphQL response but with the null fields stripped out. So the idea for the feature is to make Relay be able to understand responses with stripped nulls like this (and internally treating those as if they were null).

`markMissingAsNull` may be a bit misleading since that's not what happens in all cases (for example optimistic responses).

Anyway, don't feel strongly about any naming, will go with the one you all prefer best.

**Comment by jstejada** [comment]
I agree that `handleStrippedNulls` is also a bit confusing, and it took me a while to understand what iit even meant when I first encountered it.

I like @maraisr's suggestion, or maybe something like `treatMissingFieldsAsNull`?

> (for example optimistic responses).

@leoasis, what's the expected behavior in this case?

File: packages/relay-runtime/store/RelayModernEnvironment.js [hunk 1]
```diff
@@ -104,6 +106,7 @@ class RelayModernEnvironment implements IEnvironment {
     const handlerProvider = config.handlerProvider
       ? config.handlerProvider
       : RelayDefaultHandlerProvider;
+    this._handleStrippedNulls = config.handleStrippedNulls || false;
```
Comment (alunyov) [inline packages/relay-runtime/store/RelayModernEnvironment.js:109]: just a nit:

```suggestion
    this._handleStrippedNulls = config.handleStrippedNulls ?? false;
```

File: packages/relay-runtime/store/__tests__/RelayModernEnvironment-ExecuteWithOptimisticResponse-test.js [hunk 2]
```diff
@@ -338,4 +338,49 @@ describe('execute() with network that returns optimistic response', () => {
...
+    jest.runAllTimers();
+
+    expect(next.mock.calls.length).toBe(1);
+    expect(complete).not.toBeCalled();
+    expect(error).not.toBeCalled();
+    expect(callback.mock.calls.length).toBe(1);
+    expect(callback.mock.calls[0][0].data).toEqual({
+      me: {
+        name: 'Joe',
+      },
```
[the remark this answers is not in the document]
Comment (leoasis) [reply packages/relay-runtime/store/__tests__/RelayModernEnvironment-ExecuteWithOptimisticResponse-test.js:382]: Weird, I made the selector use the variables `{fetchSize: true}` so that it would contain the profile picture field. However I just noticed that I didn't pass that variable to the operation on execute. So I wonder why the snapshot is marked as not missing data.

[the remark this answers is not in the document]
Comment (leoasis) [reply packages/relay-runtime/store/__tests__/RelayModernEnvironment-ExecuteWithOptimisticResponse-test.js:382]: The profile picture is missing, but I just realized this is true because it only makes the selector ask for more data than the operation writes. Will change the test to make it a bit more obvious since looking up on variables defined at the top of the file makes this hard to follow.

**Comment by leoasis** [comment]
> @leoasis, what's the expected behavior in this case? I think @josephsavona mentioned that we probably don't want to apply this behavior when processing optimistic responses

Yes, currently the optimistic response case is not being filled with nulls, even when the setting is true.

However I'd like to clarify one special case, which is when the server responds with the extensions with `isOptimistic: true`. Given this is a response that comes from the server, shouldn't we fill with nulls here too?

**Comment by josephsavona** [comment]
> However I'd like to clarify one special case, which is when the server responds with the extensions with isOptimistic: true. Given this is a response that comes from the server, shouldn't we fill with nulls here too?

@kassens pointed out that sometimes a developer can't provide a reasonable optimistic value for all fields in the mutation. Instead, they may want to only provide values for the fields that they can provide a good "guess" for, and leave other fields as-is in the store until the server response is received.

I see a few options here: 
1. Make the `treatMissingFieldsAsNull` treat any field whose _value_ is `undefined` as if it were explicitly `null`.
2. Above, but special-case optimistic updates so that `treatMissingFieldsAsNull` has no effect
3. Or, make `treatMissingFieldsAsNull` treat any field that is is _missing_ as if it were explicitly `null`. This would let developers explicitly specify `undefined` as the value of keys they don't want to update - but that feels a bit too subtle.

Overall, option 2 seems best (and isOptimistic payloads would count as "optimistic", w the option disabled)

**Comment by leoasis** [comment]
> > However I'd like to clarify one special case, which is when the server responds with the extensions with isOptimistic: true. Given this is a response that comes from the server, shouldn't we fill with nulls here too?
> 
> @kassens pointed out that sometimes a developer can't provide a reasonable optimistic value for all fields in the mutation. Instead, they may want to only provide values for the fields that they can provide a good "guess" for, and leave other fields as-is in the store until the server response is received.
> 
> I see a few options here:
> 
> 1. Make the `treatMissingFieldsAsNull` treat any field whose _value_ is `undefined` as if it were explicitly `null`.
> 2. Above, but special-case optimistic updates so that `treatMissingFieldsAsNull` has no effect
> 3. Or, make `treatMissingFieldsAsNull` treat any field that is is _missing_ as if it were explicitly `null`. This would let developers explicitly specify `undefined` as the value of keys they don't want to update - but that feels a bit too subtle.
> 
> Overall, option 2 seems best (and isOptimistic payloads would count as "optimistic", w the option disabled)

Ok so option 2 is how it currently works in this PR. This also includes server sent optimistic payloads being ignored in the same way as purely client side optimistic responses.

Just double checking: if a server is stripping out the null fields from its response, if it decides to send an optimistic response it should be able to bypass that mechanism: it should explicitly send the null fields and avoid stripping them. Does this make sense? This seems to be a bit confusing for servers that implement this feature, as they are allowed to strip nulls elsewhere with the exception of this case.

**Comment by josephsavona** [comment]
@leoasis Yeah, we discussed the case of server-sent `isOptimistic` updates internally and agree with your assessment: any updates from the server should be processed using the `handleStrippedNulls` flag. In other words, the _only_ case that `handleStrippedNulls` should be ignored is when we're processing a client-supplied optimisticResponse.

@jstejada is going to import this and can help make the necessary changes just to reduce round trips (thanks again @jstejada!)

**Comment by leoasis** [comment]
@josephsavona thanks for the update! Also thanks @jstejada for making this one move forward. Is there anything else I need to do here? Not familiar with how importing this internally affects the open source repo :P

**Comment by josephsavona** [comment]
@leoasis Nothing else to do, we'll make the last few adjustments before landing (specifically, making it so that we honor the option for server-generated optimistic responses)
